README.md
This is a branch for v1.0. If you want an v0.1.2 that works with iotex-core@0.4.2 please visit [here](https://github.com/iotexproject/iotex-explorer/tree/0.1.2).

[![CircleCI](https://circleci.com/gh/iotexproject/iotex-explorer.svg?style=svg)](https://circleci.com/gh/iotexproject/iotex-explorer)

# iotex-explorer

iotex-explorer is a frontend website for [iotex-core blockchain](https://github.com/iotexproject/iotex-core).

Check our site hosted at [https://iotexscan.io](https://iotexscan.io).

Or deploy your own instance on Heroku.

<a href="https://heroku.com/deploy?template=https://github.com/iotexproject/iotex-explorer">
  <img src="https://www.herokucdn.com/deploy/button.svg" alt="Deploy">
</a>

### Development

```bash
git clone git@github.com:iotexproject/iotex-explorer.git
```

### Run your project

This is intended for \*nix users. If you use Windows, go to [Run on Windows](#run-on-windows). Let's first prepare the environment.

```bash
nvm use 10.15.0
npm install

# prepare environment variable
cp ./.env.tmpl ./.env
```

Please specify your iotex-core endpoint with the envroment variable in the `.env` file:

```
IOTEX_CORE=iotexapi.example.com:80
API_GATEWAY_URL=http://mysite.example.com/api-gateway/
```

#### Development mode

To run your project in development mode, run:

```bash
npm run watch
```

The development site will be available at [http://localhost:4004](http://localhost:4004).

#### Production Mode

It's sometimes useful to run a project in production mode, for example, to check bundle size or to debug a production-only issue. To run your project in production mode locally, run:

```bash
npm run build-production
NODE_ENV=production npm run start
```

#### NPM scripts

- `npm run test`: test the whole project and generate a test coverage
- `npm run ava ./path/to/test-file.js`: run a specific test file
- `npm run build`: build source code from `src` to `dist`
- `npm run lint`: run the linter
- `npm run kill`: kill the node server occupying the port 4004.

## Architecture

![Onefx Architecture](https://res.cloudinary.com/dohtidfqh/image/upload/v1546379050/web-guiguio/onefx-architecture.png)

```txt
.
├── README.md
├── ava.config.js           // ava test util configuration
├── babel.config.js         // babel compiler/transpiler configuration
├── config                  // project configuration
│   ├── default.js          // base config to be extended in all env
│   ├── development.js      // config in NODE_ENV=development
│   ├── production.js       // config in NODE_ENV=production
│   └── test.js             // config in NODE_ENV=test
├── coverage                // test coverage
├── dist                    // destination for src build result
├── gulpfile.babel.js       // gulp task runner config
├── package.json
├── renovate.json           // renovate bot to automate dependency bumps
├── server.ts               // project entry
├── src                               // source code
│   ├── client                        // browser-side source code
│   │   ├── javascripts
│   │   │   └── main.js
│   │   ├── static
│   │   │   ├── favicon.png
│   │   │   ├── manifest.json
│   │   │   └── robots.txt
│   │   └── stylesheets
│   │       └── main.scss
│   ├── model                         // data models
│   │   ├── index.ts
│   │   └── model.js
│   ├── server                        // onefx server
│   │   ├── babel-register.js
│   │   ├── index.ts
│   │   ├── middleware                // koa middleware
│   │   │   ├── index.ts
│   │   │   ├── manifest-middleware.js
│   │   │   └── set-middleware.js
│   │   ├── server-routes.js          // server-side routes
│   │   └── start-server.ts           // server initialization
│   └── shared                        // js code shared by both the server and the client
│       ├── app-container.js
│       ├── app.js
│       ├── common
│       ├── home
│       │   └── home.js
│       └── register-service-worker.js
├── translations          // translations supported in this website
│   ├── en.yaml
│   └── zh-cn.yaml
├── Procfile                // heroku Procfile
└── webpack.js            // webpack bundler config
```

## Run on Windows

1.  install [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10). Choose Ubuntu, for example.
2.  On WSL Ubuntu, install node version manager and install the latest lts dubnium

```bash
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.11/install.sh | bash
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion

nvm ls
nvm install lts/Dubnium
nvm use lts/dubnium
```

3.  clone repo to `C:/`

```bash
cd /mnt/c/
https://github.com/iotexproject/iotex-explorer.git
```

4.  install [VS Code](https://code.visualstudio.com/) and open WSL terminal with ctrl + \` . Not sure about WSL terminal? Go to [this post](https://blogs.msdn.microsoft.com/commandline/2017/10/27/running-node-js-on-wsl-from-visual-studio-code/).

## Contribute to Localization?

Join here to contribute: https://lokalise.co/public/903185115c91ceff0cb2a0.70666943/


CHANGELOG
2019-04-21 - 1.2.0 (1136101)
1136101 (HEAD -> master, origin/master, origin/HEAD) Update dependency gulp to v4.0.1 (#362)
830209d clean up
b2c6d3f Update dependency tslint to v5.16.0 (#345)
0a995bd Update dependency webpack to v4.30.0 (#349)
26bb91f Update npm to v6.9.0 (#351)
717b581 Update react monorepo to v16.8.6 (#352)
382867a Update dependency dotenv to v7 (#354)
e352e8a Update dependency gulp-sass to v4 (#355)
4b73fd4 Update dependency nyc to v14 (#357)
44665f7 fix console log error (#346)
a72e14b cleanup wallet
52b606b fix antd/lib/notification
f959e17 remove jsonwebtoken
b320174 continue contract interaction
365a14f Update dependency prettier to v1.17.0 (#340)
d006bfb remove unused deps
2a73d56 Update dependency @types/graphql to v14.2.0 (#324)
f998dfb Update dependency @types/dotenv to v6.1.1 (#323)
22e6dc8 add deploy contract (#328)
0907044 Update dependency @types/node-fetch to v2.3.2 (#325)
6f27b49 Update babel monorepo (#322)
7c5a80c Update Apollo GraphQL packages (#321)
bccf071 Pin dependencies (#320)
7deb4d6 import antd component not from root (#317)
6f4bbf3 add translations (#318)
e14963c add back to smart contract (#319)
9667d44 improve transfer
bbe7290 optimize dev experience / typescript is too slow (#315)
0fe66f0 fix search
5123214 update contract interact ui (#312)
4c0aa73 Finish #258 (#313)
f89e2ce bump block producers
b5dbeeb fix wallet transfer (#304)
7c9f60b fix input  border right show error (#305)
2e6c3db init pagination total number (#306)
8952f16 port solidity to abi and bytes from iotex-explorer v0.1.2 (#296)
569be8b clean up storybook
343ab7f add contract vote page (#295)
ca31932 change modal-btn styles to default by antd (#294)
95783d8 fix router push issues (#293)
e39c12c fix various issue (#292)
39c2999 add btn copy && btn link (#297)
48d57bd bump antenna (#277)
66d2067 add titles(icon, h2 title, html title) (#280)
1da22f7 fixed the box-shadow (#281)
ae37881 action detail issues (#276)
227513e add hover boxshadow change (#275)
92b575a #227 #264 sendTransfer & Transfer Result (#274)
b615c8f contract interaction success page (#273)
9ca46f9 add confirm contract modal && add horizontal-dl component && move wri… (#269)
3f70427 Merge contract & update style & fix route (#268)
ad64388 add deploy to storyboard
753d5ec move /contract -> /wallet/contract (#266)
b04d33b fix tests
2e35b94  #229 P2 1.0 Transfer Token (#260)
0469b49 Contract deploy interact (#256)
a909f04 Fix #257 Apply full width to spin outside of table so that the table overflows only within itself (#262)
c3f9c50 fix test
83df5e3 write contract step1 UI (#255)
30e0adb #217 create account with iotex-antenna (#254)
11c1cf9 Revert "#172 add an age field to bp table (#183)"
7f6e14c fix null exceptions
75ce834 Fix title (#253)
1f152c2 Merge branch 'fix-type-error' of https://github.com/uu-z/iotex-explorer
b36ef21 (uu-z-fix-type-error) change actions list count from 10 to 30 (#249)
975b84f transferAmount formRau (#248)
3fcc5a6 back home btn translation (#246)
52a10d6 Choose function (#244)
9266624 clean tp-table codes (#243)
6118b15 Receipt action (#242)
ab95332 add react storyboard
47e434d introduce iotex-react-block-producers
351a736 query  getBlockMetas by height && hash && not found - issues199 (#213)
f1ae052 #197 unlock private key with iotex-antenna (#221)
33b463f add ScrollToTop Component to app (#218)
dec03f3 if loading or error, just do not display (#219)
f3a73b6 revisit all pages and fix translations (#220)
fdefd44 Add gas price in action (#206)
fe7dfe3 bump antenna
0d34c67 #176 add wallet ui (#203)
ab965ca copy to clipboard btn improve (#211)
1139518 #173 add version information to footer (#212)
0484b21 pretty page title && sections title - Issues 202 (#214)
e5a69fd Issues 201 - replace the balance icon (#208)
afbc508 Issues 200 - address details page convert age ts to time (#207)
046afb1 Iotex core proxy (#204)
9ee360c Add a button to copy address to clipboard (#193)
1d701ab add specific title to each page (#191)
0089bfe link should not be clickable if it points to the current route (#194)
fc86134 fix block detail pagination (#189)
7b84fb1 minor tps stats
d1dc11e add coming soon page (#190)
d7ac83e #172 add an age field to bp table (#183)
e6cf9e7 fix placeholder of search box on topbar (#185)
708acbf fix types annotation by using ICtx in antenna resolver (#186)
1c44a95 redirect wallet
a83f4b1 #164 display the latest block producer and add links to go to (#167)
0315f2c add more fields to action core (#170)
9e007c9 update google analytics id (#169)
afe456e Merge pull request #171 from iotexproject/lokalise-2019-04-01_20-35-50
20f56e8 #159 improve search box (#168)
69c249f refactor getActionType
0d3aa8d #155 finish up address-details pagination (#163)
7fce429 hide mobile menu on route change (#161)
67f29a4 table responsiveness (#162)
c01be9e minor fix of transferAmount fromRau
ab1b555 Merge pull request #157 from iotexproject/lokalise-2019-03-29_00-53-56
6ab6b38 remove zh_CN
062c87a add estimate gas for action query (#149)
ab4ea36 fix: back home not working (#150)
36b01a8 table responsiveness (#151)
12d9d6f rebase master (#152)
4621e1e topbar responsiveness (#153)
f83a788 Update README.md
bb67a40 update CloudinaryImage
9f539b1 Update README.md
61b4232 udpate chain status
f257238 fix build
a6742b6 fix translations
fc80421 add comments to antenna and update doc (#145)
cb63dcb update action block render (#143)
3e20611 #111 add not found page (#144)
36ff858 fix typo
310d60d update connect-src
1f255d8 fix footer text
78cfe98 balance formatting: 4 digits precision (#141)
8d062d1 topbar responsiveness and highlight (#142)
c0860d6 add heroku integration
88989ff Update README.md
e064f92 flatten data field
b65151b update IOTEX_CORE
8211dcb bump onefx
a849138 fix fetchCoinPrice
0abeedf #88 (#97)
e93e19f bump onefx
11e420e fix z-index & dropdown scroll position (#131)
cfaa691 add SendActionResponse (#132)
ad533ff fix actHash types
0039e35 bump antenna
1b019d0 #99 (#127)
483cf24 minor fix

config/default.js
const { config } = require("dotenv");
config();

module.exports = {
  project: "iotex-explorer",
  server: {
    port: process.env.PORT || 4004,
    staticDir: "./dist",
    delayInitMiddleware: false,
    cookie: {
      secrets: ["insecure plain text", "insecure secret here"]
    },
    noSecurityHeadersRoutes: {
      "/api-gateway/": true,
      "/api/": true
    },
    noCsrfRoutes: {
      "/api-gateway/": true,
      "/api/": true,
      "/iotex-core-proxy/": true
    }
  },
  ssm: {
    enabled: false
  },
  gateways: {
    logger: {
      enabled: true,
      level: "debug"
    },
    iotexAntenna: process.env.IOTEX_CORE || "35.239.122.109:80"
  },
  analytics: {
    googleTid: "UA-111756489-2"
  },
  csp: {
    "default-src": ["none"],
    "manifest-src": ["self"],
    "style-src": [
      "self",
      "unsafe-inline",
      "https://fonts.googleapis.com/css",
      "https://use.fontawesome.com/releases/v5.0.13/"
    ],
    "frame-src": [],
    "connect-src": [
      "self",
      "https://api.coinmarketcap.com/v1/ticker/iotex/",
      "https://member.iotex.io/api-gateway/",
      process.env.API_GATEWAY_URL
    ],
    "child-src": ["self"],
    "font-src": [
      "self",
      "https://fonts.gstatic.com/",
      "https://use.fontawesome.com/releases/v5.0.13/",
      "data: https://use.fontawesome.com/releases/v5.0.13/"
    ],
    "img-src": ["*", "data:"],
    "media-src": ["self"],
    "object-src": ["self"],
    "script-src": [
      "self",
      "unsafe-eval",
      "https://www.google-analytics.com/",
      "https://ethereum.github.io/solc-bin/bin/"
    ]
  },
  apiGatewayUrl:
    process.env.API_GATEWAY_URL || "http://localhost:4004/api-gateway/",
  webBpApiGatewayUrl: "https://member.iotex.io/api-gateway/"
};
